File: ts/test/codec.test.js
/* Copyright (c) 2026 tabnas, MIT License */
'use strict'

/* codec.test.js — the escape codec, against the shared fixture. */

const { describe, it } = require('node:test')
const assert = require('node:assert')
const Path = require('node:path')

const {
  unescape, escape, findSpecDir, loadSpec, parseExpect,
} = require('../dist/support.js')


const SPEC = findSpecDir(__dirname)


describe('codec', () => {

  it('spec: util/codec.tsv', () => {
    const spec = loadSpec(Path.join(SPEC, 'util', 'codec.tsv'))
    assert.ok(0 < spec.rows.length, 'no cases')

    for (const row of spec.rows) {
      const source = row.named('source')
      const value = parseExpect(row.named('value'))
      const escaped = parseExpect(row.named('escaped'))

      assert.equal(unescape(source), value,
        `${row.where()}: unescape(${JSON.stringify(source)})`)

      assert.equal(escape(value), escaped,
        `${row.where()}: escape(${JSON.stringify(value)})`)

      // Round trip. Encoding then decoding must be the identity — that is
      // what lets a generator write a fixture the loader can read back.
      assert.equal(unescape(escape(value)), value,
        `${row.where()}: round trip`)
    }
  })

  it('leaves a cell with no backslash exactly as it is', () => {
    for (const s of ['', 'plain', 'a b c', '{"a":1}', 'ünïcödé', '𝒜𝒷']) {
      assert.equal(unescape(s), s)
      assert.equal(escape(s), s)
    }
  })

  it('decodes left to right, so escapes cannot be chained by accident', () => {
    // `\\` consumes both backslashes, leaving `n` as an ordinary letter.
    assert.equal(unescape('\\\\n'), '\\n')
    // Whereas an unescaped pair really is a newline.
    assert.equal(unescape('\\n'), '\n')
  })

  it('carries a tab and a newline through a cell', () => {
    // The two characters a TSV cell physically cannot hold.
    assert.equal(unescape('a\\tb'), 'a\tb')
    assert.equal(unescape('a\\nb'), 'a\nb')
    assert.equal(escape('a\tb'), 'a\\tb')
    assert.equal(escape('a\nb'), 'a\\nb')
  })

  it('round-trips arbitrary text', () => {
    const samples = [
      '', 'a', '\n', '\r\n', '\t\t', '\\', '\\\\', '\\n', 'a\\nb',
      'line1\nline2\r\nline3', 'C:\\dir\\file', '/^\\d+$/', '"\t"',
      'ünïcödé\t𝒜𝒷\n',
    ]
    for (const s of samples) {
      assert.equal(unescape(escape(s)), s, JSON.stringify(s))
    }
  })
})

File: rs/tests/codec_test.rs
// The escape codec, against the shared fixture. `ts/test/codec.test.js`
// and `go/escape_test.go` run the same rows.

mod common;

use tabnas_support::{escape, load_spec, parse_expect, unescape, SpecOptions, Value};

#[test]
fn spec_util_codec() {
    let spec = load_spec(
        common::spec_dir().join("util").join("codec.tsv"),
        &SpecOptions::default(),
    )
    .expect("codec.tsv loads");
    assert!(!spec.rows.is_empty(), "no cases");

    for row in &spec.rows {
        let source = row.named("source");
        let value = parse_expect(row.named("value")).expect("value is JSON");
        let escaped = parse_expect(row.named("escaped")).expect("escaped is JSON");
        let (Value::String(value), Value::String(escaped)) = (&value, &escaped) else {
            panic!("{}: value and escaped must be JSON strings", row.location());
        };

        assert_eq!(
            &unescape(source),
            value,
            "{}: unescape({source:?})",
            row.location()
        );
        assert_eq!(
            &escape(value),
            escaped,
            "{}: escape({value:?})",
            row.location()
        );
        // Round trip. Encoding then decoding must be the identity: that
        // is what lets a generator write a fixture the loader can read.
        assert_eq!(
            &unescape(&escape(value)),
            value,
            "{}: round trip",
            row.location()
        );
    }
}

#[test]
fn leaves_a_cell_with_no_backslash_exactly_as_it_is() {
    for text in ["", "plain", "a b c", "{\"a\":1}", "ünïcödé", "𝒜𝒷"] {
        assert_eq!(unescape(text), text);
        assert_eq!(escape(text), text);
    }
}

#[test]
fn decodes_left_to_right_so_escapes_cannot_be_chained_by_accident() {
    // `\\` consumes both backslashes, leaving `n` as an ordinary letter.
    assert_eq!(unescape("\\\\n"), "\\n");
    // Whereas an unescaped pair really is a newline.
    assert_eq!(unescape("\\n"), "\n");
}

#[test]
fn carries_a_tab_and_a_newline_through_a_cell() {
    assert_eq!(unescape("a\\tb"), "a\tb");
    assert_eq!(unescape("a\\nb"), "a\nb");
    assert_eq!(escape("a\tb"), "a\\tb");
    assert_eq!(escape("a\nb"), "a\\nb");
}

#[test]
fn a_trailing_lone_backslash_stays() {
    assert_eq!(unescape("back\\"), "back\\");
}
